src/metal/scheduler.ts

```diff
@@ -32,16 +32,35 @@ export class Frame implements FrameInterface {
     public timestamp: number,
     public scrollTop: number,
     public scrollLeft: number,
+    public x: number,
+    public y: number,
     public width: number,
     public height: number
   ) {}
-  static generate(): Frame {
+  static generate(root?: Element): Frame {
+    let scrollLeft = W.getScrollLeft();
```

Comment:
Instead of creating new objects and then changing them if root is defined, can you have two separate return statements for each case? It would be really nice if we didn't have to create more object allocations here since this is a hot path.